### DISCUSSION TITLE
Setting LOG_FILE_LEVEL=trace results in "Disallowed variable names in template" for _every_ template

### DISCUSSION THREAD

**Discussion by lorengordon** [renovatebot/renovate#31583] [Request Help]
### What would you like help with?

I think I found a bug

### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us which platform (GitHub, GitLab, etc) and which version of Renovate.

GitLab, Renovate version 38.89.3

### Please tell us more about your question or problem

I had some issues with private registry authentication when first setting up Renovate, and setting `LOG_LEVEL=trace` helped me understand the flow of what was happening. So I figured, as a default, "let's just capture trace logs to the log file, using `LOG_FILE_LEVEL=trace` and `LOG_FILE=renovate-log.ndjson`, it would surely be helpful for future troubleshooting!" I was still troubleshooting the private registry a little bit, and making a number of other changes to the setup, so I didn't think much of it when the logs started outputting ` INFO: Disallowed variable names in template` for _every_ template.

I eventually got the private registry authentication sorted (not a Renovate issue, just a permission thing for the token), and settled on the overall setup, and so I then started investigating this log message. I tried everything. I ripped out every change I made. Entirely back to the original setup from yesterday (just with corrected permissions on the token). No change, log still spamming this message everywhere. The very last difference, changing `LOG_FILE_LEVEL=trace` back to `LOG_FILE_LEVEL=debug`. Problem gone. 🤯

I'm _hoping_ this reproduces easily in your own environments! This is my first time setting up Renovate self-managed, so certainly open to the possibility that I screwed up _something_ that I'm just not seeing!

### Logs (if relevant)

<details><summary>Logs</summary>

```
DEBUG: Found renovate.json config file (repository=redacted/redacted-k8s)
DEBUG: Repository config (repository=redacted/redacted-k8s)
       "fileName": "renovate.json",
       "config": {
         "$schema": "https://docs.renovatebot.com/renovate-schema.json",
         "extends": ["config:recommended"],
         "ignorePaths": [],
         "ignoreDeps": ["hashicorp/terraform"],
         "packageRules": [
           {"matchManagers": ["terraform"], "groupName": "All Terraform Dependencies"}
         ]
       }
DEBUG: migrateAndValidate() (repository=redacted/redacted-k8s)
DEBUG: No config migration necessary (repository=redacted/redacted-k8s)
 INFO: Disallowed variable names in template (repository=redacted/redacted-k8s)
       "varNames": [
         null,
         "constructor",
         "allowedEnv",
         "allowCustomCrateRegistries",
         "allowedHeaders",
         "allowedPostUpgradeCommands",
         "allowPlugins",
         "allowPostUpgradeCommandTemplating",
         "allowScripts",
         "binarySource",
         "cacheDir",
         "cacheHardTtlMinutes",
         "cacheTtlOverride",
         "containerbaseDir",
         "customEnvVariables",
         "dockerChildPrefix",
         "dockerCliOptions",
         "dockerSidecarImage",
         "dockerUser",
         "dryRun",
         "encryptedWarning",
         "exposeAllEnv",
         "executionTimeout",
         "githubTokenWarn",
         "localDir",
         "migratePresets",
         "presetCachePersistence",
         "privateKey",
         "privateKeyOld",
         "gitTimeout",
         "endpoint",
         "httpCacheTtlDays",
         "autodiscoverRepoSort",
         "autodiscoverRepoOrder",
         "userAgent",
         "dockerMaxPages",
         "s3Endpoint",
         "s3PathStyle",
         "cachePrivatePackages"
       ],
       "template": "config:recommended"
```

</details>

**Comment by rarkins** [comment]
Do you mean for built-in templates too? Or just your own? How do I reproduce this? e.g. does it apply to any repo at all if we run with trace logging?

**Reply by lorengordon** [reply]
I was basing my implementation off the gitlab runner project, https://gitlab.com/renovate-bot/renovate-runner.

I believe, to reproduce, just set [`LOG_FILE_LEVEL=trace`](https://gitlab.com/renovate-bot/renovate-runner/-/blob/main/templates/renovate.gitlab-ci.yml?ref_type=heads#L9) and try to onboard a repo, or point it at any onboarded repo. The repo level config seems immaterial to the issue.

**Reply by lorengordon** [reply]
>Do you mean for built-in templates too? Or just your own?

My configs do not yet even use any templates, so I guess the answer is, yes, even the built-in templates.

Per the log above:

```
       "template": "config:recommended"
```

**Reply by Kriskras99** [reply]
I'm seeing the exact same issue. Changing `LOG_FILE_LEVEL=debug` to `LOG_FIE_LEVEL=trace` starts spamming this error:
```json
{
    "name": "renovate",
    "hostname": "...",
    "pid": 18,
    "level": 30,
    "logContext": "...",
    "repository": "...",
    "varNames": [
        null,
        "constructor",
        "allowedCommands",
        "allowedEnv",
        "allowCustomCrateRegistries",
        "allowedHeaders",
        "allowPlugins",
        "allowScripts",
        "binarySource",
        "cacheDir",
        "cacheHardTtlMinutes",
        "cacheTtlOverride",
        "containerbaseDir",
        "customEnvVariables",
        "dockerChildPrefix",
        "dockerCliOptions",
        "dockerSidecarImage",
        "dockerUser",
        "dryRun",
        "encryptedWarning",
        "exposeAllEnv",
        "executionTimeout",
        "githubTokenWarn",
        "localDir",
        "migratePresets",
        "presetCachePersistence",
        "gitTimeout",
        "endpoint",
        "httpCacheTtlDays",
        "autodiscoverRepoSort",
        "autodiscoverRepoOrder",
        "userAgent",
        "dockerMaxPages",
        "s3Endpoint",
        "s3PathStyle",
        "cachePrivatePackages"
    ],
    "template": "config:recommended",
    "msg": "Disallowed variable names in template",
    "time": "2025-08-22T11:33:20.543Z",
    "v": 0
}
```

**Comment by jamietanna** [comment]
I'd raised #38881 (a duplicate) for this, and while digging into this further, it appears that if you comment out https://github.com/renovatebot/renovate/blob/0b78d6860703212e29ae4ad0fb2dfb1473236eba/lib/util/template/index.ts#L319, no warnings appear

It's not immediately clear _why_ this happens 🤔

**Reply by jamietanna** [reply]
Using a plain `console.log` with that log line doesn't work either (on built JS):

```
function compile(template, input, filterFields = true) {
    const env = (0, utils_1.getChildEnv)({});
    const data = { ...global_1.GlobalConfig.get(), ...input, env };
    const warnVariables = new Set();
    const filteredInput = filterFields
        ? proxyCompileInput(data, warnVariables)
        : data;
    // console.log({warnVariables})
    console.log({ template, filteredInput }, 'Compiling template');
    // logger_1.logger.trace({ template, filteredInput }, 'Compiling template');
    // console.log({warnVariables})
    const result = handlebars_1.default.compile(template)(filteredInput);
    if (warnVariables.size > 0) {
        logger_1.logger.info({ varNames: Array.from(warnVariables), template }, 'Disallowed variable names in template');
    }
    return result;
}
```

There must be something that's happening when the log is called (in the underlying Bunyan logger) given:

```javascript
function compile(template, input, filterFields = true) {
    const env = (0, utils_1.getChildEnv)({});
    const data = { ...global_1.GlobalConfig.get(), ...input, env };
    const warnVariables = new Set();
    const filteredInput = filterFields
        ? proxyCompileInput(data, warnVariables)
        : data;
    console.log({warnVariables})
    // console.log({ template, filteredInput }, 'Compiling template');
    // console.loo
    logger_1.logger.trace({ template, filteredInput }, 'Compiling template');
    console.log({warnVariables})
    const result = handlebars_1.default.compile(template)(filteredInput);
    if (warnVariables.size > 0) {
        logger_1.logger.info({ varNames: Array.from(warnVariables), template }, 'Disallowed variable names in template');
    }
    return result;
}

```

Results in:

```
{ warnVariables: Set(0) {} }
TRACE ...
{
  warnVariables: Set(37) {
    Symbol(Symbol.toStringTag),
    'constructor',
    'allowedCommands',
    'allowedEnv',
    'allowCustomCrateRegistries',
    'allowedHeaders',
    'allowPlugins',
    'allowScripts',
    'binarySource',
    'cacheDir',
    'cacheHardTtlMinutes',
    'cacheTtlOverride',
    'containerbaseDir',
    'customEnvVariables',
    'dockerChildPrefix',
    'dockerCliOptions',
    'dockerSidecarImage',
    'dockerUser',
    'dryRun',
    'encryptedWarning',
    'exposeAllEnv',
    'executionTimeout',
    'githubTokenWarn',
    'localDir',
    'migratePresets',
    'presetCachePersistence',
    'gitTimeout',
    'endpoint',
    'httpCacheTtlDays',
    'autodiscoverRepoSort',
    'autodiscoverRepoOrder',
    'userAgent',
    'dockerMaxPages',
    's3Endpoint',
    's3PathStyle',
    'cachePrivatePackages',
    'ignorePrAuthor'
  }
}
```

**Reply by jamietanna** [reply]
I think it's down to how [we're creating a](https://github.com/renovatebot/renovate/blob/0b78d6860703212e29ae4ad0fb2dfb1473236eba/lib/util/template/index.ts#L316) `Proxy`, but when we log it out (at `TRACE` level) this then hits https://github.com/renovatebot/renovate/blob/0b78d6860703212e29ae4ad0fb2dfb1473236eba/lib/logger/utils.ts#L270 which then renders it, thus then triggering the warning

So the issue is that we _always_ have these warnings, we "only" see them at `TRACE`
